File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        for course, pre in prerequisites:
            graph[course].append(pre)
        
        exploring = set()
        visited = set()

        def dfs(course) -> bool:
            if course in exploring:
                return False
            
            if course in visited:
                return True
            
            exploring.add(course)

            for pre in graph[course]:
                if not dfs(pre):
                    return False
            
            exploring.remove(course)
            visited.add(course)

            return True
        
        for course, pre in prerequisites:
            if not dfs(course):
                return False
        
        return True

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<Integer>[] pre = new ArrayList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            pre[i] = new ArrayList<>();
        }

        for (int[] prerequisity: prerequisites) {
            pre[prerequisity[0]].add(prerequisity[1]);
        }

        int[] state = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (isCircle(pre, i, state)) {
                return false;
            }
        }
        return true;
    }

    private boolean isCircle(ArrayList<Integer>[] pre, int i, int[] state) {
        if (state[i] == 1) {
            return true;
        }

        if (state[i] == 2) {
            return false;
        }

        state[i] = 1;
        for (int course: pre[i]) {
            if (isCircle(pre, course, state)) {
                return true;
            }
        }
        state[i] = 2;
        return false;
    }
}